container-search/src/test/java/com/yahoo/search/significance/test/SignificanceSearcherTest.java

```diff
@@ -305,50 +303,112 @@ public void failsOnConflictingSignificanceConfiguration() {
-        Query q = new Query();
-        q.getModel().setLanguage(Language.UNKNOWN);
-        q.getRanking().setProfile("significance-ranking");
-        AndItem root = new AndItem();
-        WordItem tmp;
-        tmp = new WordItem("hello", true);
-        root.addItem(tmp);
+    // Tests that follow verify model resolving logic in different scenarios.
+    // Test naming convention is as follows:
+    // Explicit language - language set in a query
+    // Implicit language - automatically detected language
+    // Missing language - language without a model
+    // Unknown language - Language.UNKNOWN
+    // Missing word - word not in the model for specified language or fallback models
+    // Existing word - word in the model for a specified language or fallback models
 
-        q.getModel().getQueryTree().setRoot(root);
+    private Result searchWordWithLanguage(String word, Optional<Language> explicitLanguage, Optional<Language> implicitLanguage) {
+        var query = new Query();
+        explicitLanguage.ifPresent(language -> query.getModel().setLanguage(language));
+        query.getRanking().setProfile("significance-ranking");
+        var queryRoot = new AndItem();
+        var queryWord = new WordItem(word, true);
+        queryRoot.addItem(queryWord);
+        query.getModel().getQueryTree().setRoot(queryRoot);
+
+        var context = Execution.Context.createContextStub();
+        implicitLanguage.ifPresent(language -> context.setLinguistics(new MockLinguistics(language)));
+        var execution = new Execution(new Chain<>(searcher), context);
+        return execution.search(query);
+    }
 
+    private Optional<DocumentFrequency> getDocumentFrequencyWithEnglish(String word) {
         SignificanceModel model = significanceModelRegistry.getModel(Language.ENGLISH).get();
-        var helloDocumentFrequency = makeDocumentFrequency(model.documentFrequency("hello"));
-        Result r = createExecution(searcher).search(q);
+        return makeDocumentFrequency(model.documentFrequency(word));
+    }
 
-        root = (AndItem) r.getQuery().getModel().getQueryTree().getRoot();
-        WordItem w0 = (WordItem) root.getItem(0);
-        assertEquals(helloDocumentFrequency, w0.getDocumentFrequency());
+    @Test
+    public void testSignificanceSearcherWithMissingExplicitLanguageOnExistingWord() {
+        var existingWord = "hello";
+        var explicitLanguage = Language.ITALIAN;
+        var result = searchWordWithLanguage(existingWord, Optional.of(explicitLanguage), Optional.empty());
 
-        Query q2 = new Query();
-        q2.getModel().setLanguage(Language.FRENCH);
-        q2.getRanking().setProfile("significance-ranking");
-        AndItem root2 = new AndItem();
-        WordItem tmp2;
-        tmp2 = new WordItem("hello", true);
-        root2.addItem(tmp2);
+        var resultRoot = (AndItem) result.getQuery().getModel().getQueryTree().getRoot();
```

Comment:
Consider adding a helper function that extracts the DocumentFrequency from the result, to avoid these 3 lines being duplicated in all test cases.

Reply:
Makes sense. Extracted two methods and removed duplication.